### ISSUE TITLE
research: determine minimum sufficient terrain representation per Level (L4–L0)

### ISSUE THREAD

**Issue by rhythmatician** [rhythmatician/voxygen-monorepo#34]
Part of #22

## Question / Goal

> What is the minimum sufficient terrain representation at each Level L4–L0 that preserves acceptable Voxygen terrain quality while minimizing inference work, conditioning cost, data movement, and runtime latency?

This is a **decision, not an implementation**. For each Level L4..L0, determine:

1. what terrain information must be represented at that Level;
2. whether learned approximation is needed at all at that Level;
3. if learned approximation is needed, whether it should operate over:
   * 2D height/biome/scaffold information,
   * 3D NoiseRouter-derived fields,
   * a hybrid representation,
   * dense voxel/block prediction,
   * or another representation supported by evidence;
4. which existing model/runtime contracts survive **conditionally** at that Level.

Do **not** assume one representation applies to all Levels. A heterogeneous result is explicitly allowed — for example:

* L4: shortcut / scaffold
* L3: shortcut / scaffold
* L2: hybrid
* L1: dense learned residual
* L0: dense or deferred

Those are **examples of possible outcomes, not predetermined answers**. The research must decide them per Level on evidence.

Governing rule from #22: **Do not optimize an inference we can avoid. Do not optimize a class we can remove. Do not transfer data Java does not need.** This research comes *before* optimizing the dense models because it asks whether some inferences or representations can be removed entirely.

---

## Why now

The repo already exposes the seams needed for a hybrid worldgen shortcut, and has a proof that height + biome + sea level can produce a plausible semantic volume without 3D inference:

* `15 RouterField` noise, `SectionNoiseData [15,4,2,4]` native quart sampling (`RouterField.java`, `SectionNoiseData.java`, `VanillaNoiseRouterSampler.java`).
* Derived `HeightmapData [16,16]` worldSurface + oceanFloor and `BiomeProvider` abstractions.
* Fallback material painter: 8-way `SurfaceType` over top-3 layers (`HeightmapFallbackGenerator.java`).
* Semantic write boundary: `VoxelVolume(16|32)` → `VoxelVolumeWriter.writeRegion` / `VoxelVolumeWriter.writeSection` (`Level.java`, `VoxelVolume.java`, `VoxelVolumeWriter.java`).
* Existing dense control: five independent `VoxyL4..L0` U-Nets with `513×32³` heads, `ParentEncoder(513→16)` parent conditioning — captured as `docs/baselines/dense-voxy-v1.md` (**non-normative — reproducible control, not architectural authority**).

The current five-U-Net is therefore **Dense Voxy v1 — Non-Normative Baseline Snapshot**. It is the control against which shortcuts are measured. Its quality and latency fields may still be incomplete — do not turn undocumented baseline behavior into folklore.

See also `docs/MINECRAFT_NOISE_TO_VOXY_RESEARCH.md` for the full primary-source audit of the Minecraft worldgen DAG, Voxy storage, and shortcut seams.

---

## Current architectural possibilities

The frontier is deliberately left open between two poles plus the heterogeneous middle:

```
shortcut wins  →  design new contracts (scaffold → painter / residual / instances → rasterizer)
dense survives →  #24 → #25 (conditional: only where dense remains)
heterogeneous  →  e.g. L4 shortcut / L3 shortcut / L2 hybrid / L1 dense residual / L0 deferred
                 → contracts designed around that pipeline
```

Either binary outcome is useful. The ticket must keep the heterogeneous middle visible and not frame the choice as all-shortcut vs. all-dense.

### Dense Voxy v1 — what it is

Five independent models covering L4→L0, each producing a `32³` `VoxelVolume` (L4 `24×32×32` Y-trimmed) of canonical block IDs via `VoxelPredictionDecoder` → `VoxelVolumeWriter`. This is **not** the recommendation.

### Shortcut vocabulary

* **Scaffold / shortcut:** deterministic `height[32,32] + waterMask + material/band + occupancy/canopyMass` derived from `FINAL_DENSITY` zero-crossings, `SEA_LEVEL=63`, and biome — no learned 3D inference (or only lightweight 2D).
* **Hybrid:** scaffold + sparse 3D residual / instance proxies.
* **Dense residual:** scaffold + richer learned 3D correction.
* **Deferred:** no learned model at that Level (vanilla ownership or parent scaffold suffices).

---

## Experiments (no training required — prefer cheap oracle first)

All run against existing ground truth `voxy_sections(level, ws_x,y,z, labels32[32³])` joined via `>> (level+1)` bit-shift with `vocab_remap.json (1104→513)` (`build_voxy_pairs.py:108,504`, `voxy_dataset.py`). For each Level `L4..L1` (`L0` included only to verify whether learned L0 is even needed vs. vanilla ownership):

1. **Perfect height + water** — derive `height[32,32]` / `waterMask` from `FINAL_DENSITY` zero-crossing and `SEA_LEVEL=63` (and `PRELIMINARY_SURFACE_LEVEL` + aquifer fields where cheap). Reconstruct GT `32³` as `air above surface, solid below, water in between`.
2. **+ perfect surface/material** — add biome-conditioned surface band (top-3 solid layers via oracle `SurfaceType` from GT column, or GT material directly). This is the `MATERIAL PAINTER` in 2D.
3. **+ simplified feature proxies** — add oracle feature blocks via block-type set (logs/leaves/ice/mushroom etc.) either as dense mask or as `instance{type,x,z,size}` stamped by an LOD proxy (L4 canopy mass → L3 coarse mass → L2 recognizable template → L1 detailed). Measure sparse vs. dense parity.
4. **Remaining 3D residual** — `GT ⊖ reconstruction` after 1–3. This is the `sparse 3D exception field` (overhangs, arches, exposed cavities, floating shapes, externally-visible caves).
5. **Rendered-pixel impact** — same steps (1–4) but measured as screen-space error at the *intended viewing distance* for that Level (e.g. L4 silhouette at 5–20 km, L1 near-field), not just voxel argmax.
6. **Native worldgen cost vs. inference cost** — benchmark direct `NoiseRouter`/`HeightmapData` evaluation (CPU `VanillaNoiseRouterSampler` vs. GPU `GpuNoiseRouterSampler` shadow path) against ONNX per-Level latency. Answers "is learning this cheaper than a simplified vanilla eval?"
7. **Conditional information / residual entropy** — `H(residual | height, biome, y)` per Level, ideally `H(residual | previous levels' scaffold)`. Directly measures what a learned model would still need to invent.

**Shared vs. chained conditioning:** treat per-Level reconstruction two ways and compare — (a) *shared description* (each Level derived from the same underlying `height/biome/water/material` scaffold) vs. (b) *chained raster* (`L4 32³ blocks → upsample → L3 …` as in the baseline `ParentEncoder` path). Hypothesis: (a) removes serialization/upsampling/error propagation.

> Prefer cheap structural/oracle experiments first. Do not train a full new production model family merely to answer the research question unless evidence forces it.

---

## Measurements (same harness as baseline)

Use the thin harness promised in `docs/baselines/dense-voxy-v1.md` §6–7 so claims are comparable with the dense control:

* Per-stage `voxel accuracy / surface MAE / silhouette IoU / material agreement / feature recall` and the rendered-pixel metric from experiment (5).
* Per-component *cost*: conditioning (`noise`/`heightmap`/`biome`), inference (if any), decode, writer, queue latency; peak RSS / GC; `64 MiB` per `513×32³` forward reference (outer product).
* Derived "sufficiency" judgment per Level is **illustrative, not prescriptive** — e.g. `L4 → deterministic scaffold + canopy proxy (≈91–99% sufficiency)` or vindication of dense U-Nets if residual is large and vanilla eval is not cheaper.

At minimum, evaluate candidate representations **per Level** against the same world-space targets and report:

* terrain/occupancy fidelity;
* surface/height fidelity;
* semantic/block fidelity where relevant;
* topology or continuity failures;
* visible artifacts likely to matter at that Level's screen-space scale;
* estimated or measured runtime work;
* conditioning/input footprint;
* output/data-transfer footprint;
* inference count implications;
* whether the representation allows progressive refinement into the next Level.

Where `docs/research/wayfinder-research-min-sufficient-representation.md` or `docs/MINECRAFT_NOISE_TO_VOXY_RESEARCH.md` define more exact measurements, use those rather than replacing them with this summary.

---

## Per-Level exit criterion (Done when)

The research is complete only when it produces a **per-Level architecture disposition for all five Levels**. For each of L4, L3, L2, L1, L0, report:

* selected representation family (shortcut / scaffold / hybrid / dense residual / dense / deferred);
* why that representation is sufficient (numbers, not assertion);
* what information enters the representation;
* whether learned approximation exists;
* whether that learned approximation is 2D, 3D, hybrid, dense residual, or absent/deferred;
* what runtime inference remains (if any);
* what representation is handed to the next finer Level;
* whether #24 survives at that Level — `applies unchanged / applies partially / needs rewriting / does not apply`;
* whether #25 survives at that Level — same four-way;
* whether 15-field RouterField data is required at that Level (see RouterField note below);
* unresolved risks requiring prototype/training work.

The deliverable must make it possible for Wayfinder to **rewrite the #22 frontier without committing prematurely to "all dense" or "all shortcut."** The artifact is a decision markdown (committed under `docs/decisions/` or `docs/research/`) with method hashes/seeds sufficient to reproduce, ending with an explicit routing decision (shortcut vs. dense vs. heterogeneous) and the contracts to design next (or the unblocking of `#24→#25`).

---

## Dependencies / informs

* **Does NOT block #23.** The fresh-spawn `T_horizon` benchmark remains useful regardless of representation and can proceed independently.
* **Informs / gates:**
  * applicability of #24 and #25 (per-Level four-way statement required);
  * future model slimming / architecture work;
  * conditioning-cache design (Phase 4 of map);
  * RouterField contract decisions — **blocks the parked #7 RouterField single-source-generation decision** (dependency explicit, but do **not** solve #7 inside this issue);
  * scheduler cost assumptions (which Levels dominate `T_horizon`);
  * which Levels should receive expensive learned inference.

* **Feeds from (when they land):** #26 / #27 vocabulary audits — can start immediately against existing GT; their results refine the decision later but are not a gate to begin.
* **Baseline:** `docs/baselines/dense-voxy-v1.md` (control).

### RouterField — do not collapse

Do not say "RouterField is obsolete." The old 6-channel router6 conditioning design was dropped, but the **15-field Java `RouterField` / NoiseRouter boundary remains materially present** in the current codebase (`RouterField.java`, `SectionNoiseData[15,4,2,4]`). This research determines whether those 3D fields are actually part of the durable **per-Level contract**. Some Levels may require them (e.g. dense L1/L0), others may not (e.g. L4 scaffold from heightfield only).

### #24 / #25 — conditional branches

Existing tickets #24 (live L0–L4 model contracts / canonical block-ID output) and #25 (ONNX ArgMax → canonical IDs / eliminate Java logit scan) are **conditional branches, not rejections**:

* If the research says a given Level remains dense or learned in a compatible form, parts of #24/#25 may still apply there.
* If L4/L3 move to a shortcut/scaffold representation, those tickets may no longer apply uniformly across all Levels.

The per-Level statement above makes this explicit.

---

## Non-goals

* Train a full new production model family merely to answer the research question.
* Optimize ONNX kernels or perform Java micro-optimization.
* Implement #24 or #25.
* Implement RouterField single-source generation (#7).
* Decide that all Levels must share one architecture.
* Delete the dense baseline (`docs/baselines/dense-voxy-v1.md` stays as control).
* Assume visual fidelity requirements are identical at every Level.
* Turn speculative model architecture into a contract.
* Build the new `SHARED WORLDGEN SCAFFOLD → painter / residual / feature proxies → LOD-AWARE RASTERIZER` pipeline (decision only).
* Promote the dense five-U-Net interface into a durable contract.
* Require #26/#27 to finish before starting (oracle experiments run immediately).

Guardrail: distant LOD must **never** trigger vanilla chunk generation — sample deterministically via `NoiseConfig`/`DensityFunction` or shadow-router SSBOs (see `mod-optimization.md:86`, `LodGenerationService` octant-preservation path).

---

## Source artifact

This issue publishes the draft at `docs/research/wayfinder-research-min-sufficient-representation.md` (source of truth for experiments and exit criteria). That file is the `--body-file` source — keep it in sync with the issue body until closed.

Primary-source seams: `RouterField.java`, `SectionNoiseData.java`, `NoiseRouterSampler.java`, `VanillaNoiseRouterSampler.java`, `GpuNoiseRouterSampler.java`, `HeightmapData.java`, `HeightmapProvider.java`, `GpuHeightmapProvider.java`, `Level.java`, `VoxelVolume.java`, `VoxelVolumeWriter.java`, `VoxyWorldBinding.java`, `HeightmapFallbackGenerator.java`, `LodGenerationService.java`, `voxy_models.py`, `voxy_export.py`, `build_voxy_pairs.py`, `voxy_dataset.py`, `CanonicalRegistries.java`, `external/voxy/.../Mapper.java`. Full audit in `docs/MINECRAFT_NOISE_TO_VOXY_RESEARCH.md`.

Type: `wayfinder:research`

> **Frontier note (blocking semantics):** cheap oracle/scaffold experiments that support `worldgen-partition` may proceed immediately (this research informs #85); final **interpretation/closure** of this ticket waits on #85's disposition matrix. This avoids idle research while keeping the GitHub `blocked_by` edge meaningful for the decision.

**Comment by rhythmatician** [comment]
Reopened because the stated research exit criterion is not satisfied. The closure comment published a decision template but explicitly left experiments 1-7, the common measurement-harness results, and the eleven-field L4-L0 decision table to future work. Publication of the specification is not completion of the measured decision. Continue this existing #22 child; do not create a duplicate or apply agent:implement while architectural research remains unresolved.

**Comment by rhythmatician** [comment]
> **Upstream synthesis note (2026-08-16):** This research is now a **downstream consumer** of minecraft-worldgen-decomposition (#82), terrain-signal-lattices (#83), correct-distant-terrain (#84), worldgen-partition (#85), and refinement-topology (#86). It should **not independently answer** DAG, lattice/phase, partition, topology, or 'correct' definitions — it synthesizes those evidences into per-Level representation dispositions (shortcut / hybrid / dense residual / deferred) plus conditional survival of #24/#25 and RouterField per-Level requirements. See #22 Lane A/B/C (map restructured). This note does not change the existing exit criteria; it scopes the synthesis dependency. The map body now carries the expanded lanes; #34 remains the synthesis ticket.

**Comment by rhythmatician** [comment]
### Update — per-Level sufficiency oracle experiments (no training needed)

**Decision scope (#34, wayfinder:research):** per Level L4–L0 determine representation family (shortcut/scaffold/hybrid/dense residual/dense/deferred), why sufficient (numbers), information in, learned 2D/3D/hybrid/absent, inference left, hand to next, #24/#25 four-way survival, 15-field need, risks. Heterogeneous allowed (`L4 scaffold / L3 scaffold / L2 hybrid / L1 dense / L0 deferred` are examples). Baseline is **Dense Voxy v1** (`VoxyL4..L0` 513×32³ U-Nets, `ParentEncoder 513→16`) — non-normative control, keep it, measure shortcuts vs it.

**7 cheap oracle experiments** on GT `voxy_sections(level→labels32)` via `>>(level+1)` + `vocab_remap 1104→513`: 1 height+water from `FINAL_DENSITY` zero-cross+`SEA_LEVEL=63`, 2 +surface/material (2D `SurfaceType` top-3 oracle), 3 +feature proxies (`instance{type,x,z,size}` L4 mass→L3→L2→L1), 4 `GT⊖recon` residual (overhangs/arches/caves), 5 rendered-pixel at Level viewing distance (L4 5–20 km), 6 native vs inference cost (`VanillaNoiseRouterSampler` vs `GpuNoiseRouterSampler`), 7 `H(residual|height,biome,y)`. Compare shared scaffold vs chained raster (`L4→upsample→L3` ParentEncoder) — shared should remove error propagation. Measure `voxel accuracy/surface MAE/silhouette IoU/material/feature recall` + `conditioning/inference/decode/writer/queue + RSS/GC + 64MiB/513×32³` and per-result sufficiency to route Wayfinder rewrite of #22 frontier. Guardrail: no Level triggers vanilla chunk gen — use `NoiseConfig/DensityFunction`/SSBO.

Source: `docs/research/wayfinder-research-min-sufficient-representation.md` + `RouterField.java`, `SectionNoiseData`, `HeightmapData`, `VoxelVolumeWriter`, `HeightmapFallbackGenerator.SurfaceType`.

**Comment by rhythmatician** [comment]
## Minimum-representation evidence from PR #218 (End only)

The merged End tracer is a strong qualitative result for this research:

- L4→L2 require no learned model in the current End slice; seeded vanilla density plus Level-specific deterministic rasterization produced terrain the maintainer judged worth preserving.
- L1 can also avoid learned inference for base terrain by using an exact sampler-direct vanilla noise walk.
- The semantic representation is deliberately minimal: `air | end_stone`; placed features and structures are outside the claim.

This supports a heterogeneous/no-model disposition for End base geometry, but it is not the quantitative per-Level exit evidence #34 requires. The run is one seed/profile and lacks the metric suite/cost comparison. #219 is especially important counter-evidence: L4 coarse occupancy can over-connect islands and substantially reshape during refinement.

Do not generalize this result to Overworld/Nether or to feature/material responsibilities.
